servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/Host.java

```diff
@@ -0,0 +1,491 @@
+                    // Connection was already removed, nothing to do.
+                    break;
+                } else if (connections.length == 1) {
+                    assert !Host.isUnhealthy(currentConnState) : "Cannot be UNHEALTHY with #connections > 0";
+                    if (Host.isActive(currentConnState)) {
+                        if (connStateUpdater.compareAndSet(this, currentConnState,
+                                new ConnState(EMPTY_ARRAY, currentConnState.state))) {
+                            break;
+                        }
+                    } else if (currentConnState.state == State.EXPIRED
+                            // We're closing the last connection, close the Host.
+                            // Closing the host will trigger the Host's onClose method, which will remove the host
+                            // from used hosts list. If a race condition appears and a new connection was added
+                            // in the meantime, that would mean the host is available again and the CAS operation
+                            // will allow for determining that. It will prevent closing the Host and will only
+                            // remove the connection (previously considered as the last one) from the array
+                            // in the next iteration.
+                            && connStateUpdater.compareAndSet(this, currentConnState, CLOSED_CONN_STATE)) {
+                        this.closeAsync().subscribe();
+                        break;
+                    }
+                } else {
+                    Object[] newList = new Object[connections.length - 1];
+                    System.arraycopy(connections, 0, newList, 0, i);
+                    System.arraycopy(connections, i + 1, newList, i, newList.length - i);
+                    if (connStateUpdater.compareAndSet(this,
+                            currentConnState, new ConnState(newList, currentConnState.state))) {
+                        break;
+                    }
+                }
+            }
+            LOGGER.trace("{}: removed connection {} from {} after {} attempt(s).",
+                    lbDescription, connection, this, removeAttempt);
+        }).onErrorComplete(t -> {
+            // Use onErrorComplete instead of whenOnError to avoid double logging of an error inside subscribe():
+            // SimpleCompletableSubscriber.
+            LOGGER.error("{}: unexpected error while processing connection.onClose() for {}.",
+                    lbDescription, connection, t);
+            return true;
+        }).subscribe();
+        return true;
+    }
+
+    // Used for testing only
+    @SuppressWarnings("unchecked")
+    Map.Entry<Addr, List<C>> asEntry() {
+        return new AbstractMap.SimpleImmutableEntry<>(address,
+                Stream.of(connState.connections).map(conn -> (C) conn).collect(toList()));
+    }
+
+    public Object[] connections() {
```

Comment:
Even though the whole class is pkg-private, consider keeping methods pkg-private as well